pkg/util/pull/pull.go

```diff
@@ -15,14 +15,113 @@ import (
 
 type Getter = func(context.Context) error
 
-// OrPoll will attempt to call pull, falling back to poll if it's not supported.
-// pull subscribe to changes, blocking until the subscription fails, if possible failures will be retried.
-// poll should query for the current state, this function manages delays between calls.
-func OrPoll(ctx context.Context, pull, poll Getter, opts ...Option) error {
+// OrPoll calls pull, falling back to repeated calls to get if pull returns a codes.Unimplemented error.
+// Argument pull should subscribe to changes, blocking until the subscription fails.
+// Argument get should get the current state, blocking until the state is got or an error occurs.
+// OrPoll blocks until the context is cancelled, retrying either pull or get on error.
+func OrPoll(ctx context.Context, pull, get Getter, opts ...Option) error {
 	conf := calcOpts(opts...)
 
-	shouldPoll := false
+	return runBlocking(ctx, "blocking tasks", func(ctx context.Context) error {
+		err := runPull(ctx, pull, conf)
+		if shouldReturn(err) {
+			return err
+		}
+		// the runPull err is never nil, so if we're here then we need to try polling
+		err = get(ctx) // get one to see if this is unimplemented too
+		if shouldReturn(err) {
+			return err
+		}
+		if !isUnimplemented(err) {
+			// pull is unimplemented but get is not, there's a chance things have changed (i.e. during boot)
+			// so try pull again just to be sure.
+			pullErr := runPull(ctx, pull, conf)
+			if shouldReturn(pullErr) {
+				return pullErr
+			}
+			// pull is still unimplemented, let's start the poll loop for real
+			err = runPoll(ctx, err, get, conf)
+			if shouldReturn(err) {
+				return err
+			}
+			// else get is unimplemented
+		}
+
+		// if we're here both pull and get returned unimplemented, so retry everything after a delay
+		return err
+	}, conf, func(error) bool { return false })
+}
+
+// Changes calls Pull on poller unless it's not supported, in which case it polls.
+// It will retry on error, backing off exponentially up to a maximum delay.
+// It will return if the context is cancelled or a non-recoverable error occurs.
+func Changes[C any](ctx context.Context, poller Fetcher[C], changes chan<- C, opts ...Option) error {
+	return OrPoll(ctx,
+		func(ctx context.Context) error {
+			return poller.Pull(ctx, changes)
+		},
+		func(ctx context.Context) error {
+			return poller.Poll(ctx, changes)
+		},
+		opts...)
+}
+
+// runPull retries calling pull returning on context cancellation or when pull returns an error that has codes.Unimplemented.
+// runPull will never return a nil error.
+func runPull(ctx context.Context, pull Getter, conf changeOpts) error {
+	return runBlocking(ctx, "pulls", pull, conf, isUnimplemented)
+}
+
+// runPoll blocks repeatedly calling get after some delay until context is cancelled or get returns an error that is not codes.Unimplemented.
```

Comment:
Comment says that runPoll returns when context is cancelled or a non-Unimplemented error is returned by get. However, the function will return when an Unimplemented error happens (line 108). It returns when the error is a context error or codes.Unimplemented, other errors will retry.

Reply:
Ooops, that'd be a typo then